### DISCUSSION TITLE
Reduce the noise, work, and frustration from CVE reporting

### DISCUSSION THREAD

**Discussion by wesleytodd** [npm/feedback#62] [CLI]
**Work In Progress**: I will be making updates to this top level post in response to the discussions below.

## Problem

Today a small group of maintainers get bombarded with issues and dependency update PRs when a new CVE is reported on popular libraries. Many, if not most, of these are false positives from a vulnerability perspective. This level of noise creates distrust between security companies/researchers, maintainers, and end users. It also frustrates users and maintainers due to the shear volume of work it creates for them. In a worst case scenario, this leads to even worse security practices; in best case, it catches some vulnerabilities at the cost of ongoing maintenance work.

There are many examples of this noise leading to frustrations and work for maintainers and end users, here are a few example links:

- https://github.com/facebook/create-react-app/issues/8529
- https://github.com/handlebars-lang/handlebars.js/issues/1668
- https://github.com/conventional-changelog/conventional-changelog/issues/592 (this one was not called out as a false positive, but it is)

## Proposal

Introduce a new feature in the CLI which will load an "ignore list" which will filter out audit results before displaying to end users.  The source of this list should be configurable such that the work of maintaining up-to-date lists can be federated across the ecosystem.

### Goals

- Reduce the workload for maintainers and end users when dealing with CVE reports
- Improve security hygiene
- Move the investigation and decision making about the applicability of a CVE to those closest to the issue
- Achieve some alignment among popular tools on how to handle this problem
- Allow enterprises and companies to align their interests with the OSS community while also supporting it

### Prior Art

This proposal is an evolution of work by @naugtur in the [Audit Resolver tooling](https://www.npmjs.com/package/npm-audit-resolver).  There are many previous conversations: 

- https://github.com/nodejs/package-maintenance/issues/386
- https://github.com/npm/rfcs/pull/18
- https://github.com/npm/cli/pull/10
- https://npm.community/t/interactive-tool-to-manage-audit-findings-npm-audit-resolve/197

### Technical Details

*TBD*:  We think that this problem is more universal than just npm or even JavaScript projects. To ensure that we come to a solution which will work across all JS reporting tools we are setting up an OpenJS Collab Space to bring people across the ecosystem together.  We will hopefully come out of that with technical details which can be implemented by npm, snyk, yarn, and others.

*Note:* Between the work on the audit resolver and discussions in the Package Maintenance WG we have a working idea, but before we officially write it up I wanted to start this part of the meta discussion.

## Next Steps

- [ ] Setup [Collab Space](https://github.com/openjs-foundation/cross-project-council/blob/master/COLLABORATION_NETWORK.md)
- [ ] Consolidate discussions into one digestible write-up (maybe in the collab space?)

**Comment by darcyclarke** [comment]
@wesleytodd thanks for filing this! I know we've had several discussions on this topic already & are going to continue this conversation in the OpenJS Collab Space. 

I just want to make sure we're on the same page - which we might not be & that's totally okay; We've talked on & off about this for over a year now & my memory isn't what it used to be. I think the goals you've outlined all make sense although I'm not sold on the solution being a federated ignore list (I know we've ideated on it before).

IMO, real change will require something more fundamental. I've brought this up before - & recently discussed internally - creating a mechanism to resolve or counterclaim CVEs at their source (ex. counters: *"this isn't valid because of x"* or *"this isn't valid for my **dependent package** because of y"*); Then, that mutable resolution list would live alongside the vuln DB itself (in the case of `npm` that would likely be a new &/or update to the existing security endpoints in the registry which third-parties & tooling could match or support). I think this would go further to address the root cause which is a lack of check, balance & nuance in how CVEs are currently filed & ultimately resolved in the ecosystem (which, as you know, involves significant churn re: patch/publish/updateᴺ). Today, it's a one-sided story - Tomorrow, it could be a conversation.

> - Reduce the workload for maintainers and end users when dealing with CVE reports
> - Improve security hygiene
> - Move the investigation and decision making about the applicability of a CVE to those closest to the issue
> - Achieve some alignment among popular tools on how to handle this problem
> - Allow enterprises and companies to align their interests with the OSS community while also supporting it

I think these goals, specifically, highlight why we should do our best to align with & improve the existing mechanics. As a maintainer, fragmenting/federating sounds as if I'll actually have a harder time getting end-users/consumers sync'd up & resolving noise if ignore lists become decentralized.

Definitely looking forward to continuing this discussion though & hope we can help improve the situation at-large.

**Note:** I might be okay with supporting a root-level resolution/ignore list if it wasn't configured via `package.json`, ignored during `publish` & not dependent on a third-party/network - ie. could potentially live in `.npmrc` config since it would be mapped against/tied to the `registry` you're using (the [overrides](https://github.com/npm/rfcs/pull/129/files#diff-321aff6a180122b0908ba16a5f20e873f3a4d007048baad34de20defce81039a) spec is a good example - highlighting that we're keen on supporting these type of features moving forward).

**Reply by isaacs** [reply]
> As a maintainer, fragmenting/federating sounds as if I'll actually have a harder time getting end-users/consumers sync'd up & resolving noise if ignore lists become decentralized.

Worth noting that this is already a problem with fragmented vendor-specific advisory lists.  I've gotten pings about false positives that were corrected in npm's vulnerability list because they still exist in snyk's.  (I'm sure the inverse probably happens as well, I just have more ability to correct npm's list when these things happen, for obvious reasons 😅)

**Reply by wesleytodd** [reply]
> I just want to make sure we're on the same page - which we might not be & that's totally okay;

Yep, the goal is to get on the same page.  It is probably better if we all start with different ideas!

> IMO, real change will require something more fundamental.

I would *love* to see something change at a more fundamental level. During our last video call on this topic I thought most folks agreed that moving the entire ship was so difficult it might not succeed.  The more modest proposal for the ignore list was something we settled on, but if we think we can actually change the whole system then I am more than happy to have that conversation.

>  I might be okay with supporting a root-level resolution/ignore list if it wasn't configured via package.json

I purposely left the technical details out so we didn't go too far down this rabbit hole here, but I completely agree we cannot have a `package.json` based solution.

> Worth noting that this is already a problem with fragmented vendor-specific advisory lists.

Completely agree.  I think the problem is that putting together a list is something many people can do easily, supporting the tooling necessary to apply that list is the missing link here.

**Comment by lumaxis** [comment]
Just leaving a quick note here since the original post also mentions automatic update PRs: GitHub just made a few changes to the Dependabot feature https://github.blog/changelog/2020-10-27-a-quieter-dependabot/

**Comment by mayakacz** [comment]
Jumping in. Disclosure: I work at GitHub on software supply chain security.

> Goals
> * Reduce the workload for maintainers and end users when dealing with CVE reports
> * Improve security hygiene
> * Move the investigation and decision making about the applicability of a CVE to those closest to the issue
> * Achieve some alignment among popular tools on how to handle this problem
> * Allow enterprises and companies to align their interests with the OSS community while also supporting it

If I had to try to summarize this, what you're ultimately getting at is "don't bother me with vulnerabilities that don't affect me". For notifications, that could be a response in npm audit, it could be a notification or alert; and for the environment, you could be referring to vulnerabilities that aren't exploitable, don't apply in that environment, aren't severe enough to warrant attention, etc.
I think the important part to note here is that it's *specific* to that user. Each user will have a different environment, and different preferences for notifications.

In an ideal world, for a vulnerability in a dependency you have, you would be able to tell and change your behaviour based on:
* Its severity (Low/ Moderate/ High/ Critical)
* Whether a fix exists
* Whether an exploit exists
* Whether the vulnerable path is being used in your environment
* Whether the dependency applies to your environment (e.g., devDependency for prod)
* Whether the vulnerability applies to your environment (e.g., a DDoS vuln for a firewalled environment, or a developer toolchain exploit for a non-dev environment)
etc.

Practically, this information will need to be automatically generated, not provided by a user or maintainer, as that would be too time consuming for every vulnerability and dependency. There's also a potential for 'lying' or mistakes if a maintainer provides this information for their package (though unlikely, I see that as a particularly attractive entrypoint for an attacker if it existed).

> Introduce a new feature in the CLI which will load an "ignore list" which will filter out audit results before displaying to end users. The source of this list should be configurable such that the work of maintaining up-to-date lists can be federated across the ecosystem.

So then, considering the above, I don't think this is the 'right' solution. I think incremental changes towards the above (which many products, including GitHub's, are doing), and the ability to transitively pass this information down the supply chain to other dependencies (via a Software Bill of Materials, say), is a more flexible and sustainable way to tackle this problem, not dependent on a single source of truth.

Now unfortunately there isn't a perfect solution for this today (as I mentioned, these steps are incremental), but I believe this is where the industry is heading. Would like your thoughts.

**Reply by ljharb** [reply]
I'm not sure how, in dynamic languages, this kind of thing *can* be reliably automatically generated.

Specifically, I'm not saying "don't bother *me*", i'm asking for a way to "do not bother my *users* with things that do not affect my package". The only possible way I can conceive of to determine that is if i, the package's maintainer, have a way to manually claim that this is the case.

The potential for lying or mistakes imo is far far less risky than what's currently happening: the potential for consumers to turn off or learn to ignore all security notifications entirely.

**Reply by mayakacz** [reply]
> I'm not sure how, in dynamic languages, this kind of thing can be reliably automatically generated.

Some of the above can.

> Specifically, I'm not saying "don't bother me", i'm asking for a way to "do not bother my users with things that do not affect my package". The only possible way I can conceive of to determine that is if i, the package's maintainer, have a way to manually claim that this is the case.

I understand your use case. I'm saying that what's missing here is a way to reliably generate this information (which we may disagree on), and separately, a *way to reliably pass on this information*. That's what something like SBOM efforts are trying to address. You should be able to consume that information for all of your dependencies, even for transitive dependencies several layers deep. You should be able to specify not just "doesn't apply" but "doesn't apply to dev environments" or something more specific than just yes/no. And you should be able to consume this information in a machine-readable way, to quickly and easily make decisions.

**Reply by ljharb** [reply]
I don't have a formed opinion about the way to pass on the information, and everything you're saying about that seems great!

It seems useful to have multiple kinds of information to decide to pass on, and for users to decide to trust - why would shared lists of "doesn't apply"s, that users can opt in to, not be a useful piece of information to the ecosystem?

**Reply by MylesBorins** [reply]
@mayakacz it would appear that the solution you are advocating for, which I find very compelling, is one where the Vulnerability report itself would have more granularity + signal about the risk of a vulnerability and that module consumers would then in turn be able to filter or silence warnings based on policy.

One thing I am unclear on is if such a policy would only exist at the top level of an application, or if a module author could implement such a policy in transitive dependencies. I'm assuming that folks will likely want to be able to pass a flag to see all potential warnings... but it is likely that a module author has the most context about which of these signals are relevant for their modules... but they also shouldn't be making the decision for a consumer to silence something they may care about.

**Comment by meblum** [comment]
High level overview of this issue. https://overreacted.io/npm-audit-broken-by-design/
